Web Inspector: Implicitly nested property declarations inside non-style rules results in nested content being deleted during editing or displaying incorrect matched styles for elements

patrickangle opened the pull request:
#### b1dd655b710e45f59c91ef803255a8dbc4fc94e1
<pre>
Web Inspector: Implicitly nested property declarations inside non-style rules results in nested content being deleted during editing or displaying incorrect matched styles for elements
<a href="https://bugs.webkit.org/show_bug.cgi?id=251565">https://bugs.webkit.org/show_bug.cgi?id=251565</a>
rdar://104821946

Reviewed by Devin Rousso.

InspectorStyleSheet was built with the assumption that Style rules contained property declarations, and nothing else.
CSS nesting has proven this assumption wrong in significant ways by allowing properties and other rules to be declared,
interleaved inside a rule. This includes inside @ rules, which previously could not contain property delcarations
directly.

This means during editing, we only want to replace the property declarations, not the full body of a rule, since the
rule may contain other rules. Luckily, canonically all property declarations occur before all nested rules, so we can
safely move nested rules below to make our lives a bit easier with no harm to the meaning of the style sheet.

This fix adds instrumentation to the CSS parser so we can be informed of the new &quot;implicit&quot; nested rule that contains
properties inside of non-style rules. It also overhauls how rule body text is edited, much like we had to last year for
rule header text.

* LayoutTests/inspector/css/getMatchedStylesForNode-expected.txt:
- Account for new property of CSSRule.

* LayoutTests/inspector/css/getMatchedStylesForNodeNestingStyleGrouping-expected.txt:
* LayoutTests/inspector/css/getMatchedStylesForNodeNestingStyleGrouping.html:
* LayoutTests/inspector/css/modify-css-property-expected.txt:
* LayoutTests/inspector/css/modify-css-property.html:
* LayoutTests/inspector/css/resources/modify-css-property.css:
- Add test cases for implicitly nested rules and their siblings, children, and parents.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
- Mark implicitly nested rules so that the frontend can prevent the editing of their selector.

* Source/WebCore/css/CSSPropertySourceData.h:
(WebCore::CSSRuleSourceData::CSSRuleSourceData):
- The container rule types can now contain properties, so we always need to have the buffer for that information ready,
since we won&apos;t be informed by the parser that an implicit nested context was created until after we have observed the
properties themselves.

* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeRegularRuleList):
(WebCore::CSSParserImpl::consumeDeclarationListOrStyleBlockHelper):
(WebCore::CSSParserImpl::consumeStyleBlock):
* Source/WebCore/css/parser/CSSParserImpl.h:
- Don&apos;t send duplicate bodyStart/bodyEnd messages to the observer.
- Notify the observer when the engine has created an implict nested rule inside a body.

* Source/WebCore/css/parser/CSSParserObserver.h:

* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::atRuleIdentifierForType):
(WebCore::isValidRuleHeaderText):
- Pull the mapping of types to their keyword text for reuse in setting new style text.

(WebCore::StyleSheetHandler::endRuleBody):
- In order to maintain parity with CSSOM&apos;s representation of styles, we need to create an implictly nested rule to
match against the CSSOM&apos;s implicitly nested rule. This also allows us to inform the frontend that said CSSOM rule was
implicitly nested, since the OM itself doesn&apos;t carry this information.

(WebCore::StyleSheetHandler::markRuleBodyContainsImplicitlyNestedProperties):
- Observe to mark the style rule data as containing implicitly nested properties, which will then trigger us to take those properties and mvoe them to a special implicit style rule data object.

(WebCore::InspectorStyleSheet::buildObjectForRule):

(WebCore::isNotSpaceOrTab):
(WebCore::InspectorStyleSheet::setRuleStyleText):
- Overhaul the setting of style text.
- Undo/oldText is handled specially because we want to restore the sheet back to the non-canonical form after an edit.
- The frontend does not provide nested rules as part of its new text, so we must readd them ourselves, which we can do
from the original style sheet.
- Indentation is matched to the new property delcartion text provided by the frontend.

(WebCore::InspectorStyleSheetForInlineStyle::setRuleStyleText):
- Match the method signature for non-inline styles, and simplify the indirection previously present for setting a style
rule&apos;s/inline style&apos;s text.

(WebCore::InspectorStyle::setText): Deleted.
(WebCore::InspectorStyleSheet::setStyleText): Deleted.
(WebCore::InspectorStyleSheetForInlineStyle::setStyleText): Deleted.
* Source/WebCore/inspector/InspectorStyleSheet.h:

* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
- Adopt new method to set the body text of a rule.

* Source/WebInspectorUI/UserInterface/Models/CSSRule.js:
(WI.CSSRule):
(WI.CSSRule.prototype.get isImplicitlyNested):
* Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
(WI.CSSStyleDeclaration.prototype.get selectorEditable):
* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles.prototype._parseRulePayload):
- Mark the selector of implicitly nested rules as non-editable.

Canonical link: <a href="https://commits.webkit.org/261329@main">https://commits.webkit.org/261329@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ee4354830c5410955a9ea536a5fa478718bc7e2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111169 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20308 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43729 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2593 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/120005 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115122 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21682 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11409 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/2216 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116925 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16107 "Found 1 new test failure: fast/scrolling/keyboard-scrolling-distance-pageDown.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99286 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103729 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98064 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30948 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44622 "") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/99737 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12823 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/32283 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/86510 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state-changed, /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/event-listener (failure)") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/10925 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13359 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/9275 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/100889 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18781 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51873 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31458 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7850 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15310 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/108924 "Built successfully") | 
| | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/26860 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -1436,23 +1471,90 @@ Ref<Protocol::CSS::CSSStyle> InspectorStyleSheet::buildObjectForStyle(CSSStyleDe
+        return { };
+    }
+
+    auto indentation = emptyString();
+    auto startOfSecondLine = newText.find('\n');
+    if (startOfSecondLine != notFound) {
+        ++startOfSecondLine;
+        auto endOfSecondLineWhitespace = newText.find(isNotSpaceOrTab, startOfSecondLine);
+        if (endOfSecondLineWhitespace != notFound)
+            indentation = newText.substring(startOfSecondLine, endOfSecondLineWhitespace - startOfSecondLine);
+    }
```
I'm hopeful someone has a better way to accomplish this and the indentation groking below 😞

dcrousso replied:
i wonder if we should maybe change `Inspector::ContentSearchUtilities::countRegularExpressionMatches` to instead be `getRegularExpressionMatches` and find the most common (or average) line whitespace prefix

patrickangle replied:
Because we get the properties pre-formatted from the frontend, the average of everything but the first and last line should always be the same as just getting the indentation of a single line. The last line is then some fraction of that average, and the first line should always have no space (since the "text" is everything between the `{}`.

mdubet commented on Source/WebCore/css/CSSPropertySourceData.h:
```diff
@@ -105,14 +104,15 @@ struct CSSRuleSourceData : public RefCounted<CSSRuleSourceData> {
-    // Only for CSSMediaRules.
     RuleSourceDataList childRules;
+
+    bool isImplicitlyNested = false;
```
We generally use brace-initialization in WebCore  `bool isImplicitlyNested { false }`

mdubet commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -125,6 +125,23 @@ static CSSParserContext parserContextForDocument(Document* document)
+        return "@container"_s;
+    default:
+        ASSERT_NOT_REACHED();
+            return ""_s;
```
Weird indentation

mdubet commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -582,10 +582,16 @@ Vector<RefPtr<StyleRuleBase>> CSSParserImpl::consumeRegularRuleList(CSSParserTok
     if (isNestedContext()) {
         runInNewNestingContext([&]() {
-            consumeStyleBlock(block, StyleRuleType::Style);
+            static const auto isParsingStyleDeclarationsInRuleList = true;
```
What is the goal of this const variable instead of directly passing `true` to the consumeStyleBlock function ?

dcrousso replied:
maybe have an `enum class ParsingStyleDeclarationsInRuleList : bool { No, Yes }` instead?

mdubet commented on Source/WebCore/inspector/InspectorStyleSheet.h:
```diff
@@ -241,6 +239,7 @@ class InspectorStyleSheetForInlineStyle final : public InspectorStyleSheet {
     void didModifyElementAttribute();
     ExceptionOr<String> text() const final;
     CSSStyleDeclaration* styleForId(const InspectorCSSId& id) const final { ASSERT_UNUSED(id, !id.ordinal()); return &inlineStyle(); }
+    ExceptionOr<void> setRuleStyleText(const InspectorCSSId&, const String& newText, String* oldText, bool isUndo = false) final;
```
Maybe use an `enum class IsUndo { Yes, No }` instead of a raw boolean

mdubet commented on Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
```diff
@@ -165,12 +165,13 @@ class InspectorCSSAgent::SetStyleTextAction final : public InspectorCSSAgent::St
     ExceptionOr<void> undo() override
     {
-        return m_styleSheet->setStyleText(m_cssId, m_oldText, nullptr);
+        static const auto isUndo = true;
```
Same concern about this variable

dcrousso commented on LayoutTests/inspector/css/getMatchedStylesForNodeNestingStyleGrouping.html:
```diff
@@ -112,6 +147,16 @@
         color: red;
 
+        @media(min-width: 0px) {
+            background-color: blue;
```
is this actually useful/needed?

patrickangle replied:
This line mostly exists so that the test exercises the potentially less common path of the property we want being below an explicitly nested rule.

dcrousso commented on LayoutTests/inspector/css/getMatchedStylesForNodeNestingStyleGrouping.html:
```diff
@@ -112,6 +147,16 @@
+                color: yellow;
+            }
+
+            color: purple;
```
ooohhh so `isImplicitlyNested` covers cases like this (i.e. the same selector, but a different `@rule`)?  i wonder if maybe we could name it something like `isNestedAtRule` to clarify that instead 🤔

patrickangle replied:
isNestedAtRule is less clear IMO since the styles are not in an @rule according to CSSOM, but are instead implicitly nested as a style rule inside the @rule.

dcrousso commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -582,10 +582,16 @@ Vector<RefPtr<StyleRuleBase>> CSSParserImpl::consumeRegularRuleList(CSSParserTok
+                // property declarations.
                 rules.append(createNestingParentRule());
+
+                if (m_observerWrapper)
+                    m_observerWrapper->observer().markRuleBodyContainsImplicitlyNestedProperties();
```
it's moments like these i wish C++ had optional chaining 🥲

dcrousso commented on Source/WebCore/css/parser/CSSParserImpl.cpp:
```diff
@@ -1182,9 +1190,9 @@ void CSSParserImpl::consumeDeclarationList(CSSParserTokenRange range, StyleRuleT
 }
 
-void CSSParserImpl::consumeStyleBlock(CSSParserTokenRange range, StyleRuleType ruleType)
+void CSSParserImpl::consumeStyleBlock(CSSParserTokenRange range, StyleRuleType ruleType, bool isParsingStyleDeclarationsInRuleList)
```
ditto

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -350,12 +365,34 @@ void StyleSheetHandler::endRuleBody(unsigned offset)
     auto rule = popRuleData();
     fixUnparsedPropertyRanges(rule.ptr());
+
+    if (rule->containsImplicitlyNestedProperties&& rule->styleSourceData->propertyData.size()) {
```
Style: missing space before `&&`

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -350,12 +365,34 @@ void StyleSheetHandler::endRuleBody(unsigned offset)
+        std::swap(rule->styleSourceData, inferredStyleRule->styleSourceData);
+
+        // Inferred style rules are always logically placed at the start of their parent rule.
+        rule->childRules.insert(0, WTFMove(inferredStyleRule));
```
are there WPT for this that we previously would've failed but now pass?

patrickangle replied:
WPT does not test inspector behavior. We previously would end up with a mismatch between the CSSOM and our flattened source data vector, which meant that Web Inspector would show the incorrect rules for these, and likely for styles that followed in the style sheet as well.

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -1403,6 +1435,9 @@ RefPtr<Protocol::CSS::CSSRule> InspectorStyleSheet::buildObjectForRule(CSSStyleR
     if (groupingsPayload->length())
         result->setGroupings(WTFMove(groupingsPayload));
 
+    if (auto&& sourceData = ruleSourceDataFor(rule))
```
does this actually return a `&&`?  a bit odd for us to explicitly say that without actually `WTFMove` or anything

patrickangle replied:
This appears to be a typo on my part. I can't figure out a reason I would have written this 😅

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -1436,23 +1471,90 @@ Ref<Protocol::CSS::CSSStyle> InspectorStyleSheet::buildObjectForStyle(CSSStyleDe
 }
 
-ExceptionOr<void> InspectorStyleSheet::setStyleText(const InspectorCSSId& id, const String& text, String* oldText)
+static inline bool isNotSpaceOrTab(UChar character)
```
NIT: i kinda feel like this should be flipped (i.e. `isSpaceOrTab`) and the callsites do a `!` as that'd be easier to read (and likely more composable with other checks if needed in the future)

patrickangle replied:
That doesn't quite work since this is a function we pass to String::find to find the first character that isn't a space or tab...

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -1436,23 +1471,90 @@ Ref<Protocol::CSS::CSSStyle> InspectorStyleSheet::buildObjectForStyle(CSSStyleDe
+
+ExceptionOr<void> InspectorStyleSheet::setRuleStyleText(const InspectorCSSId& id, const String& newText, String* oldText, bool isUndo)
+{
+    RefPtr<CSSRule> cssRule = ruleForId(id);
```
`auto`?

dcrousso commented on Source/WebInspectorUI/UserInterface/Models/CSSRule.js:
```diff
@@ -25,7 +25,7 @@
 WI.CSSRule = class CSSRule extends WI.Object
 {
-    constructor(nodeStyles, ownerStyleSheet, id, type, sourceCodeLocation, selectorText, selectors, matchedSelectorIndices, style, groupings)
+    constructor(nodeStyles, ownerStyleSheet, id, type, sourceCodeLocation, selectorText, selectors, matchedSelectorIndices, style, groupings, isImplicitlyNested)
```
Aside: we should eventually move all the non-required stuff into an `{...} = {}` argument

dcrousso commented on Source/WebCore/inspector/InspectorStyleSheet.cpp:
```diff
@@ -1436,23 +1471,90 @@ Ref<Protocol::CSS::CSSStyle> InspectorStyleSheet::buildObjectForStyle(CSSStyleDe
+    }
+
+    auto indentation = emptyString();
+    auto startOfSecondLine = newText.find('\n');
```
does CSS allow for `\r` as a newline instead?

patrickangle replied:
Maybe, but here the input text always comes from the Web Inspector frontend where the string being built has `\n` used in CSSStyleDeclaration.js

burg reviewed (commented):
Nit in changelog: 'declarations'

patrickangle committed:
Web Inspector: Implicitly nested property declarations inside non-style rules results in nested content being deleted during editing or displaying incorrect matched styles for elements
https://bugs.webkit.org/show_bug.cgi?id=251565
rdar://104821946

Reviewed by Devin Rousso.

InspectorStyleSheet was built with the assumption that Style rules contained property declarations, and nothing else.
CSS nesting has proven this assumption wrong in significant ways by allowing properties and other rules to be declared,
interleaved inside a rule. This includes inside @ rules, which previously could not contain property delcarations
directly.

This means during editing, we only want to replace the property declarations, not the full body of a rule, since the
rule may contain other rules. Luckily, canonically all property declarations occur before all nested rules, so we can
safely move nested rules below to make our lives a bit easier with no harm to the meaning of the style sheet.

This fix adds instrumentation to the CSS parser so we can be informed of the new "implicit" nested rule that contains
properties inside of non-style rules. It also overhauls how rule body text is edited, much like we had to last year for
rule header text.

* LayoutTests/inspector/css/getMatchedStylesForNode-expected.txt:
- Account for new property of CSSRule.

* LayoutTests/inspector/css/getMatchedStylesForNodeNestingStyleGrouping-expected.txt:
* LayoutTests/inspector/css/getMatchedStylesForNodeNestingStyleGrouping.html:
* LayoutTests/inspector/css/modify-css-property-expected.txt:
* LayoutTests/inspector/css/modify-css-property.html:
* LayoutTests/inspector/css/resources/modify-css-property.css:
- Add test cases for implicitly nested rules and their siblings, children, and parents.

* Source/JavaScriptCore/inspector/protocol/CSS.json:
- Mark implicitly nested rules so that the frontend can prevent the editing of their selector.

* Source/WebCore/css/CSSPropertySourceData.h:
(WebCore::CSSRuleSourceData::CSSRuleSourceData):
- The container rule types can now contain properties, so we always need to have the buffer for that information ready,
since we won't be informed by the parser that an implicit nested context was created until after we have observed the
properties themselves.

* Source/WebCore/css/parser/CSSParserImpl.cpp:
(WebCore::CSSParserImpl::consumeRegularRuleList):
(WebCore::CSSParserImpl::consumeDeclarationListOrStyleBlockHelper):
(WebCore::CSSParserImpl::consumeStyleBlock):
* Source/WebCore/css/parser/CSSParserImpl.h:
- Don't send duplicate bodyStart/bodyEnd messages to the observer.
- Notify the observer when the engine has created an implict nested rule inside a body.

* Source/WebCore/css/parser/CSSParserObserver.h:

* Source/WebCore/inspector/InspectorStyleSheet.cpp:
(WebCore::atRuleIdentifierForType):
(WebCore::isValidRuleHeaderText):
- Pull the mapping of types to their keyword text for reuse in setting new style text.

(WebCore::StyleSheetHandler::endRuleBody):
- In order to maintain parity with CSSOM's representation of styles, we need to create an implictly nested rule to
match against the CSSOM's implicitly nested rule. This also allows us to inform the frontend that said CSSOM rule was
implicitly nested, since the OM itself doesn't carry this information.

(WebCore::StyleSheetHandler::markRuleBodyContainsImplicitlyNestedProperties):
- Observe to mark the style rule data as containing implicitly nested properties, which will then trigger us to take those properties and mvoe them to a special implicit style rule data object.

(WebCore::InspectorStyleSheet::buildObjectForRule):

(WebCore::isNotSpaceOrTab):
(WebCore::InspectorStyleSheet::setRuleStyleText):
- Overhaul the setting of style text.
- Undo/oldText is handled specially because we want to restore the sheet back to the non-canonical form after an edit.
- The frontend does not provide nested rules as part of its new text, so we must readd them ourselves, which we can do
from the original style sheet.
- Indentation is matched to the new property delcartion text provided by the frontend.

(WebCore::InspectorStyleSheetForInlineStyle::setRuleStyleText):
- Match the method signature for non-inline styles, and simplify the indirection previously present for setting a style
rule's/inline style's text.

(WebCore::InspectorStyle::setText): Deleted.
(WebCore::InspectorStyleSheet::setStyleText): Deleted.
(WebCore::InspectorStyleSheetForInlineStyle::setStyleText): Deleted.
* Source/WebCore/inspector/InspectorStyleSheet.h:

* Source/WebCore/inspector/agents/InspectorCSSAgent.cpp:
- Adopt new method to set the body text of a rule.

* Source/WebInspectorUI/UserInterface/Models/CSSRule.js:
(WI.CSSRule):
(WI.CSSRule.prototype.get isImplicitlyNested):
* Source/WebInspectorUI/UserInterface/Models/CSSStyleDeclaration.js:
(WI.CSSStyleDeclaration.prototype.get selectorEditable):
* Source/WebInspectorUI/UserInterface/Models/DOMNodeStyles.js:
(WI.DOMNodeStyles.prototype._parseRulePayload):
- Mark the selector of implicitly nested rules as non-editable.

Canonical link: https://commits.webkit.org/261329@main
